File: 22.括号生成.cpp
/*
 * @lc app=leetcode.cn id=22 lang=cpp
 * @lcpr version=30203
 *
 * [22] 括号生成
 */

// @lc code=start
class Solution {
public:
    vector<string> generateParenthesis(int n) {
        backtrack(0, n);
        return res;
    }

private:
    vector<string> res;
    string track = "";
    int left = 0, right = 0;
    void backtrack(int idx, int n) {
        if (left < right) {
            return;
        }
        if (idx > 2 * n) {
            return;
        }
        if (left + right == 2 * n && left == right) {
            res.push_back(track);
            return;
        }

        left++;
        track += '(';
        backtrack(idx + 1, n);
        left--;
        track.pop_back();

        right++;
        track += ')';
        backtrack(idx + 1, n);
        right--;
        track.pop_back();
    }
};
// @lc code=end



/*
// @lcpr case=start
// 3\n
// @lcpr case=end

// @lcpr case=start
// 1\n
// @lcpr case=end

 */

File: 22.括号生成.java
/*
 * @lc app=leetcode.cn id=22 lang=java
 * @lcpr version=30305
 *
 * [22] 括号生成
 */

// @lc code=start
class Solution {
    List<String> res = new ArrayList<>();
    public List<String> generateParenthesis(int n) {
        trackback(n, 0, 0, new StringBuilder());
        return res;
    }

    private void trackback(int n, int left, int right, StringBuilder track) {
        if (right > left || left > n || right > n) {
            return;
        }
        if (left == n && right == n) {
            res.add(track.toString());
            return;
        }
        
        track.append('(');
        trackback(n, left + 1, right, track);
        track.deleteCharAt(track.length() - 1);

        track.append(')');
        trackback(n, left, right + 1, track);
        track.deleteCharAt(track.length() - 1);

    }
}
// @lc code=end



/*
// @lcpr case=start
// 3\n
// @lcpr case=end

// @lcpr case=start
// 1\n
// @lcpr case=end

 */

